doc/generators.md

```diff
@@ -20,25 +20,34 @@ The two main object types are `directory` and `generator`. The root of `generato
 **Directory objects** take the following keys:
 * `type`: This must be set to `directory`.
 * `testdata.yaml`: Optional yaml configuration that will be copied to `testdata.yaml` in this directory.
-* `solution`: Optional invocation of a solution to be used to generate `.ans` files. Set to empty to disable generating `.ans`. (Useful for e.g. the `data/samples/` directory.)
-    * Relative paths are relative to `generators/`.
-    * Absolute paths are relative to the problem root.
+* `solution`: Optional invocation of a solution to be used to generate `.ans` files. Set to empty to disable generating `.ans`. (Useful for e.g. the `data/samples/` directory.) This must be an absolute path relative to the problem root.
 * `visualizer`: Optional invocation of a visualizer to generate visualizations for each test case in this directory.
-    * Relative paths are relative to `generators/`.
-    * Absolute paths are relative to the problem root.
+    This must be an absolute path relative to the problem root. Set to empty to disable.
 * `random_salt`: Optional string that will be prepended to each command before computing its `{seed}`. May be used to regenerate all random cases and to prevent predictable seeds.
 * `data`: The test cases / test groups contained in this directory. This may take two forms:
     * A dictionary, each key is the name of a test case/test group, and each value must be a `directory` or `generator` object.
-    * A list of dictionaries as above. In this case, testcases will automatically be prefixed with numbers in the order of the list.
-* `generators`: The root directory may contain a dictionary listing generators and their dependencies. This may be used when using non-directory programs as generators which depend on other files in the `generators/` directory. Each key of the dictionary is the name of a generator, and values must be lists of file paths relative to `generators/`.
+    * A list of dictionaries as above. In this case, testcases will be prefixed with zero padded 1-based integers in the order of the list. Items in the same dictionary will get the same number.
 
 **Generator objects** have one of the following forms:
-- `null`: An empty generator means that the testcase is a manual case and must not be modified or deleted by generator tooling.
+- `null`: An empty generator means that the testcase is a manual case and must not be modified or deleted by generator tooling. The corresponding `.in` file must be present in the `data/` directory. The corresponding `.ans` may be present, but may also be generated from a given solution. Note that this form is discouraged. Prefer specifying a path to a `.in` file as below.
```

Comment:
I would add something like `but may also be generated *once* from a given solution` to clarify that it won't be regenerated.